### ISSUE TITLE
Bug report - Model consistently predicts confidence of above 90% on samples with no speech

### ISSUE THREAD

**Issue by greerviau** [snakers4/silero-vad#309]
## 🐛 Bug

The model seems to consistently predicts confidence of above 90% on samples with no speech. 
Edit: Testing again it appears that it does predict 0% on true silence, but any background noise causes it to predict speech.

## To Reproduce

Im running a custom listener that uses Silero-VAD

samplerate = 16000

```python
  class SileroVADListener(Listener):
      def __init__(self, wake_word: str, device_idx: int, samplerate: int, sensitivity: int):
          super().__init__(wake_word, device_idx, samplerate)
  
          self.model, self.utils = torch.hub.load(repo_or_dir='snakers4/silero-vad',
                                model='silero_vad',
                                force_reload=False)
  
          self.blocksize = int(self.samplerate / 10)
  
          self.speech_detected_thresh = 3
  
      # Taken from utils_vad.py
      def validate(self, model, inputs: torch.Tensor):
          with torch.no_grad():
              outs = model(inputs)
          return outs
  
      # Provided by Alexander Veysov
      def int2float(self, sound):
          abs_max = np.abs(sound).max()
          sound = sound.astype('float32')
          if abs_max > 0:
              sound *= 1/abs_max
          sound = sound.squeeze()  # depends on the use case
          return sound
      
      def is_speech(self, audio_chunk):
          
          audio_int16 = np.frombuffer(audio_chunk, np.int16)
  
          audio_float32 = self.int2float(audio_int16)
          
          # get the confidences and add them to the list to plot them later
          new_confidence = self.model(torch.from_numpy(audio_float32), 16000).item()
          print(new_confidence)
          return True
  
      def listen(self, engaged: bool):
  
          # Define a flag to indicate if speech has started
          recording_started = False
  
          # Set up the sounddevice stream
          stream = pyaudio.PyAudio().open(
              format=pyaudio.paInt16,
              channels=1,
              rate=self.samplerate,
              input=True,
              frames_per_buffer=self.blocksize)
  
          speech_detected = 0
  
          while True:
              try:
                  # Get audio frames from the buffer
                  frame = stream.read(1536, exception_on_overflow=False)
  
                  if self.is_speech(frame):
                      print('speech')
                      speech_detected += 1
                  else:
                      print('no speech')
                      speech_detected = 0
  
                  if speech_detected > self.speech_detected_thresh and not recording_started:
                      print('Speech detected')
                      print('Recording...')
                      recording_started = True
                      self.reset_recording()
                      # Write the recording buffer to the file
                      for f in self.recording_buffer:
                          self.record_frame(f)
                  elif not speech_detected and recording_started:
                      print('Recording stopped')
                      recording_started = False
                      self.stop_recording()
                      # Clear out the recording buffer
                      self.recording_buffer.clear()
                      #return self.get_audio_data()
                      
                  # If we have detected the hotword start writing audio frames to file
                  # If we havent detected the hotword, write audio frames to the buffer
                  # The recording buffer avoids a cutoff at the begining of the recording
                  if recording_started:
                      self.record_frame(frame)
                  else:
                      self.recording_buffer.append(frame)
  
              except KeyboardInterrupt:
                  break
```

An example of output confidence on silence
`0.9915153980255127
speech
0.9888885617256165
speech
0.9892972111701965
speech
0.9887569546699524
speech
0.9844241738319397
speech
0.9854767918586731
speech
0.9891263842582703
speech
0.988828182220459
speech
0.9919223785400391
speech
0.9928280711174011
speech
0.9896829724311829
speech
0.9890046119689941
speech`

## Expected behavior

I would expect it to have a confidence below 50%

## Environment

certifi            2022.9.24
cffi               1.15.1
charset-normalizer 2.1.1
click              8.1.3
Flask              2.2.2
idna               3.4
importlib-metadata 5.0.0
itsdangerous       2.1.2
Jinja2             3.1.2
MarkupSafe         2.1.1
numpy              1.23.4
pip                22.0.4
playsound          1.3.0
PyAudio            0.2.12
pycparser          2.21
pydub              0.25.1
requests           2.28.1
scapy              2.4.5
setuptools         58.1.0
simpleaudio        1.0.4
sounddevice        0.4.5
soundfile          0.12.1
SpeechRecognition  3.8.1
srt                3.5.2
torch              1.13.1
torchaudio         0.13.1
tqdm               4.64.1
typing_extensions  4.5.0
urllib3            1.26.12
uuid               1.30
vosk               0.3.44
webrtcvad          2.0.10
websockets         10.4
Werkzeug           2.2.2
zipp               3.9.0

## Additional context

I am testing this on a macbook air, although I wouldnt expect that to have an effect. I have also tested webrtc-vad in a similar listener class, and I also notice it predicts speech on silence as well. This makes me believe that there potentially is an underlying issue that I am not seeing. I will try testing on another machine to verify.

**Comment by snakers4** [comment]
Please send your audio for us to test.

**Comment by snakers4** [comment]
Also, there is a proper iterator class in the utils as an example, your code for example does not reset model state on start.

**Comment by adamnsandle** [comment]
This code snippet
```python3
model, utils = torch.hub.load(repo_or_dir='snakers4/silero-vad',
                                                model='silero_vad',
                                                force_reload=True)

(get_speech_timestamps,
 save_audio,
 read_audio,
 VADIterator,
 collect_chunks) = utils

wav = read_audio('files/recording.wav')
raw_probs = model.audio_forward(wav, 16000, 512)
```
produces following result:
![image](https://user-images.githubusercontent.com/36505480/225590903-3960ba4f-dd72-44a9-880d-0968485d645d.png)
`.audio_forward` just splits audio into small chunks and iterates over them (like in your streaming examle)

maybe something wrong with audio normalization during preprocessing

**Comment by greerviau** [comment]
I followed the pyaudio stream example, has the method for normalizing the audio frames changed from that example?

`def int2float(self, sound):
        abs_max = np.abs(sound).max()
        sound = sound.astype('float32')
        if abs_max > 0:
            sound *= 1/abs_max
        sound = sound.squeeze()  # depends on the use case
        return sound
    
    def is_speech(self, audio_chunk):
        
        audio_int16 = np.frombuffer(audio_chunk, np.int16)

        audio_float32 = self.int2float(audio_int16)
        
        # get the confidences and add them to the list to plot them later
        confidence = self.model(torch.from_numpy(audio_float32), 16000).item()
        print(confidence)`

This is the code I am using to normalize and make predictions, pulled from that example. The input to `is_speech` is just the audio frame pulled from the pyaudio stream.

**Comment by adamnsandle** [comment]
Replacing 
```python3
if abs_max > 0:
    sound *= 1/abs_max
```
with
```python3
if abs_max > 0:
    sound *= 1/32768
```
in `int2float` might help

**Comment by snakers4** [comment]
Please open a new issue, if this will be relevant in future.
